Populate machine status with libvirt domain details

bison opened the pull request:
This populates the machine's status subresource with the libvirt domain's details.

bison commented:
/hold

Still a WIP. Almost there.

frobware commented on cloud/libvirt/actuators/machine/actuator.go:
```diff
@@ -62,11 +92,125 @@ func (a *Actuator) Delete(cluster *clusterv1.Cluster, machine *clusterv1.Machine
+	libvirtStatus, err := libvirtutils.ProviderStatus(machine)
+	if err != nil {
+		glog.Errorf("Could not get provider status for update: %v", err)
+		return err
```
Why do we, in general, log errors, then return an err? Isn't the caller going to log the error, and perhaps with additional context?

bison replied:
Mostly doing this in the "top-level" Update / Create / etc. methods because they are the boundary between the actuator and the upstream controller. I've run into cases where the controller didn't log the returned errors which caused a lot of wasted time debugging.

frobware commented on cloud/libvirt/actuators/machine/utils/domain.go:
```diff
@@ -745,3 +754,125 @@ func DomainExists(machine *clusterv1.Machine) (bool, error) {
+		return nil, fmt.Errorf(LibVirtConIsNil)
+	}
+
+	dom, err := virConn.LookupDomainByUUID(uuid[:])
```
Having got this far... I'm wondering why we pass a pointer to the uuid? Is it because it's an underlying C/native type?

bison replied:
Ended up dropping the UUID thing altogether and just looking up by name. It's simpler and should be fine.

frobware commented on cloud/libvirt/actuators/machine/utils/domain.go:
```diff
@@ -745,3 +754,125 @@ func DomainExists(machine *clusterv1.Machine) (bool, error) {
+
+	status, ok := obj.(*providerconfigv1.LibvirtMachineProviderStatus)
+	if !ok {
+		return nil, fmt.Errorf("unexpected object: %#v", gvk)
```
Would expressing `%T` be more useful here? i.e., what type did we get?

bison replied:
This `gvk` struct will have the group, version, and kind information. Honestly, this bit is just lifted from the AWS actuator, but it seems like a reasonable thing to log in this situation.

frobware commented on cloud/libvirt/actuators/machine/utils/domain.go:
```diff
@@ -745,3 +754,125 @@ func DomainExists(machine *clusterv1.Machine) (bool, error) {
+		return "Shutoff"
+	}
+
+	return "Unknown"
```
Make this the `default` case?

bison replied:
Sure. Done.

bison commented on cloud/libvirt/actuators/machine/actuator.go:
```diff
@@ -44,12 +57,29 @@ func NewActuator(params ActuatorParams) (*Actuator, error) {
+	dom, err := libvirtutils.LookupDomainByUUID(machine, uuid)
+	if err != nil {
+		glog.Errorf("Could not lookup created libvirt machine: %v", err)
+		return fmt.Errorf("error looking up created machine %v", err)
```
Fixed.

bison commented on cloud/libvirt/actuators/machine/actuator.go:
```diff
@@ -44,12 +57,29 @@ func NewActuator(params ActuatorParams) (*Actuator, error) {
+
+	if err := a.updateStatus(machine, dom); err != nil {
+		glog.Errorf("Could not update machine status: %v", err)
+		return fmt.Errorf("error updating machine status: %v", err)
```
Fixed.

bison commented:
/hold cancel

bison commented:
I think this is ready.

ingvagabund commented on cloud/libvirt/actuators/machine/actuator.go:
```diff
@@ -62,11 +105,139 @@ func (a *Actuator) Delete(cluster *clusterv1.Cluster, machine *clusterv1.Machine
+		return error.WithLog(err, "error getting provider status")
+	}
+
+	uuid, err := uuid.Parse(*libvirtStatus.InstanceID)
```
This panics when `libvirtStatus.InstanceID` is `nil`. In case you call the `Update` from scratch without populating machine status.

ingvagabund replied:
It should be fine to look based on domain name (to be define under `cloud/libvirt/actuators/machine/utils/domain.go`):
```go
func LookupDomainByName(machine *clusterv1.Machine) (*libvirt.Domain, error) {
	machineProviderConfig, err := MachineProviderConfigFromClusterAPIMachineSpec(&machine.Spec)
	client, err := buildClient(machineProviderConfig.URI)
	if err != nil {
		return nil, fmt.Errorf("Failed to build libvirt client: %s", err)
	}

	virConn := client.libvirt
	if virConn == nil {
		return nil, fmt.Errorf(LibVirtConIsNil)
	}

	dom, err := virConn.LookupDomainByName(machine.Name)
	if err != nil {
		return nil, fmt.Errorf("Failed to lookup domain by name: %v", err)
	}

	return dom, nil
}
```

bison replied:
I ditched the whole UUID thing. It didn't make sense to look up the domain via a field stored in the status in order to update the status anyway. The domain is looked up by name now.

ingvagabund commented on cloud/libvirt/actuators/machine/actuator.go:
```diff
@@ -62,11 +105,139 @@ func (a *Actuator) Delete(cluster *clusterv1.Cluster, machine *clusterv1.Machine
 func (a *Actuator) Update(cluster *clusterv1.Cluster, machine *clusterv1.Machine) error {
 	glog.Infof("Updating machine %v for cluster %v.", machine.Name, cluster.Name)
-	return fmt.Errorf("TODO: Not yet implemented")
+	error := errorWrapper{cluster: cluster, machine: machine}
```
Maybe `errorWr`? I got confused by the `error.WithLog`. The `error` is allocated for `error` type. We should not override it unless necessary.

bison replied:
Good point. Fixed.

paulfantom commented on Dockerfile:
```diff
@@ -15,6 +15,8 @@
 # Reproducible builder image
 FROM openshift/origin-release:golang-1.10 as builder
 
+RUN yum install -y libvirt-devel
```
Do not add this as `libvirt-devel` is already in `openshift/origin-release:golang-1.10` image

bison replied:
Looks like this was already fixed in master. This bit is gone after a rebase.

enxebre commented on cloud/libvirt/actuators/machine/utils/domain.go:
```diff
@@ -757,3 +766,104 @@ func DomainExists(machine *clusterv1.Machine) (bool, error) {
+
+// UpdateProviderStatus updates the provider status in-place with info
+// from the given libvirt domain.
+func UpdateProviderStatus(status *providerconfigv1.LibvirtMachineProviderStatus, dom *libvirt.Domain) error {
```
any machine api specific logic does not belong here, we can address it after merging and rebasing https://github.com/openshift/cluster-api-provider-libvirt/pull/38

bison replied:
Fixed.

frobware commented on cloud/libvirt/actuators/machine/utils/domain.go:
```diff
@@ -551,35 +554,35 @@ func domainDefInit(domainDef *libvirtxml.Domain, name string, machineConfig prov
+func createDomain(name string, machineProviderConfig *providerconfigv1.LibvirtMachineProviderConfig, offset int) (*uuid.UUID, error) {
 	if name == "" {
-		return fmt.Errorf("Failed to create domain, name is empty")
+		return nil, fmt.Errorf("Failed to create domain, name is empty")
```
I don't think this is 'failed to create domain'. Rather, 'name is empty'. :-/

frobware commented on cloud/libvirt/actuators/machine/utils/domain.go:
```diff
@@ -551,35 +554,35 @@ func domainDefInit(domainDef *libvirtxml.Domain, name string, machineConfig prov
 	// Get values from machineProviderConfig
 	if err := domainDefInit(&domainDef, name, *machineProviderConfig); err != nil {
-		return fmt.Errorf("Failed to init domain definition from machineProviderConfig: %v", err)
+		return nil, fmt.Errorf("Failed to init domain definition from machineProviderConfig: %v", err)
```
We should consistently use `%v` for error. In the few lines above we use `%s`.

frobware commented on cloud/libvirt/actuators/machine/utils/domain.go:
```diff
@@ -551,35 +554,35 @@ func domainDefInit(domainDef *libvirtxml.Domain, name string, machineConfig prov
 		}
 	} else {
-		return fmt.Errorf("machine does not has a IgnKey value")
+		return nil, fmt.Errorf("machine does not has a IgnKey value")
```
typo: `machine does not _have_ a IgnKey value`.

bison commented on cloud/libvirt/actuators/machine/actuator.go:
```diff
@@ -44,12 +72,27 @@ func NewActuator(params ActuatorParams) (*Actuator, error) {
+
+	uuid, err := libvirtutils.CreateVolumeAndMachine(machine, a.cidrOffset)
+	if err != nil {
+		return error.WithLog(err, "error creating libvirt machine")
```
Sure, but that's upstream and this means we can actually debug now.

Anyway, this was logging the error via the wrapper. The naming is just confusing. I changed it as suggested in another comment.

bison commented on cloud/libvirt/actuators/machine/actuator.go:
```diff
@@ -62,11 +105,139 @@ func (a *Actuator) Delete(cluster *clusterv1.Cluster, machine *clusterv1.Machine
+// object.
+func EncodeLibvirtMachineProviderStatus(status *providerconfigv1.LibvirtMachineProviderStatus) (*runtime.RawExtension, error) {
+	// TODO: Modifying this in place feels weird. Should it be copied?
+	status.TypeMeta = metav1.TypeMeta{
```
This is what the AWS actuator does, but yeah, it felt weird. I don't think it's needed if we just set the metadata upfront when we first create a status, so I've done that instead and removed this bit.

bison commented:
Rebased and simplified this quite a bit. I think it's probably ready as-is, but a few open questions:

1. The status field names are copied from the AWS actuator. Should we update them to make more sense here? e.g. `DomainUUID` instead of `InstanceID`, or is there a compelling reason to have them match the AWS status?

2. I haven't included any conditions in the status. Happy to do that now if we think it's important, or can tackle it in a follow up PR.

paulfantom commented:
/approve
/hold

frobware commented:
> 2\. I haven't included any conditions in the status. Happy to do that now if we think it's important, or can tackle it in a follow up PR.

I vote w do that in a follow-up PR. Better to keep them small and easily reviewable (and revertable).
